atomix/cluster/src/test/java/io/atomix/cluster/messaging/impl/NettyMessagingServiceTest.java

```diff
@@ -592,99 +587,107 @@ public void testCompletableRemoteHandlerFailureEmptyStringValue() {
-    final ManagedMessagingService nettyWithOwnPool =
-        createMessagingServiceWithSpiedPool(channelsOpen, channelRef);
+    final MutableReference<ChannelPool> poolRef = new MutableReference<>();
+    final var timeoutOnCreate = Duration.ofSeconds(10);
+    final var nettyWithOwnPool = createMessagingServiceWithPool(poolRef);
+    final var channelPool = poolRef.get();
 
-    netty2.registerHandler(
-        subject, (address, bytes) -> new CompletableFuture<>()); // never complete this future
-    final CompletableFuture<byte[]> response =
-        nettyWithOwnPool.sendAndReceive(address2, subject, "fail".getBytes());
-    assertThat(response)
-        .failsWithin(Duration.ofSeconds(15))
-        .withThrowableOfType(ExecutionException.class)
-        .havingRootCause()
-        .isInstanceOf(expectedException.getClass())
-        .withMessageContaining("timed out in");
-
-    // when
-    verify(channelRef.get(), timeout(1000).atLeast(1)).close();
-    nettyWithOwnPool.sendAndReceive(address2, subject, "fail".getBytes());
-
-    // then
-    // closing causes an CloseException which causes another close
-    Awaitility.await("channels should be recreated").until(channelsOpen::get, c -> c == 2);
+    try {
+      nettyWithOwnPool.start().join();
+
+      // grab the original channel, so we can assert it was closed
+      // it's also useful to send a successful request once, so we can ensure the channel exists and
+      // set a lower timeout on the request we want specifically to time out
+      netty2.registerHandler(subject, (address, bytes) -> new byte[0], Runnable::run);
+      nettyWithOwnPool
+          .sendAndReceive(address2, subject, "get channel".getBytes(), true, timeoutOnCreate)
+          .join();
+      final var originalChannel = channelPool.getChannel(address2, subject).join();
+
+      // set up handler which will always cause timeouts
+      netty2.unregisterHandler(subject);
+      netty2.registerHandler(subject, (address, bytes) -> new CompletableFuture<>());
+      final CompletableFuture<byte[]> response =
+          nettyWithOwnPool.sendAndReceive(
+              address2, subject, "fail".getBytes(), true, Duration.ofSeconds(1));
+      assertThat(response)
+          .failsWithin(Duration.ofSeconds(15))
+          .withThrowableThat()
+          .havingRootCause()
+          .isInstanceOf(TimeoutException.class);
+
+      // when - remote connection finally succeeds
+      // give a generous time out on the second request, as creating a new channel can be slow at
+      // times
+      netty2.unregisterHandler(subject);
+      netty2.registerHandler(subject, (address, bytes) -> new byte[0], Runnable::run);
+      nettyWithOwnPool.sendAndReceive(
+          address2, subject, "success".getBytes(), true, timeoutOnCreate);
+
+      // then
+      final var newChannel = channelPool.getChannel(address2, subject).join();
+      assertThat(originalChannel.closeFuture()).succeedsWithin(Duration.ofSeconds(15));
```

Comment:
:x: The test is still flaky. In the local run it failed in 292nd run.

When the test was flaky (before this PR), I observed that there is a delay in closing `originalChannel`. So the last `sendAndReceive` sometimes gets the old channel, instead of creating the new one. I think this is just a timing issue. So I think it helps to move this line before calling the last `sendAndReceive`.

Reply:
Good catch :+1: I did run it in a loop but it never failed on my laptop, so I can't really say if that will fix everything, but it sounds likely